Dockerfile COPY instead of git clone

m1k1o opened the pull request:
Changed dockerfile so that it uses data from context directory and does not need to clone git repository. Also reduced number of RUN tags, because each creates a new layer.

m1k1o committed:
dockerfile copy

Maddosaurus commented:
Oh, it has been ages since I last looked at this Dockerfile, and it is in dire need of some love - thank you so much for that PR!
Much needed changes that optimize this quite a lot!

Maddosaurus commented on Dockerfile:
```diff
@@ -1,12 +1,17 @@
 USER obscura
-CMD python3 main.py
+
+ENTRYPOINT [ "python3" ]
```
Nitpick: I know python3 is in the path, but it might be more robust, to set it to its full path here (just checked against a current alpine: `/usr/bin/python3`), correct? That eliminates one more source of errors. What do you think?
Also: This combination of ENTRYPOINT/CMD is a great idea, thanks!

m1k1o replied:
I agree, it could be more explicit. Or maybe even better solution would be to switch to `FROM python:3-alpine` image, with python already installed.

Maddosaurus replied:
Let's have a look at that together. I haven't seem the project in quite some time and I am unsure what happens if we switch the base image. @cmllr do you see any problems there?

m1k1o commented:
With python base image it could be like this (+ with caching requirements.txt and all dependencies).

```Dockerfile
FROM python:3-alpine

WORKDIR /cameraobscura

COPY requirements.txt requirements.txt

RUN apk add --no-cache build-base jpeg-dev zlib-dev shadow; \
    LIBRARY_PATH=/lib:/usr/lib python3 -m pip install -r requirements.txt; \
    groupadd -r obscura && useradd -r -g obscura -d /cameraobscura -s /sbin/nologin -c "Docker image user" obscura; \
    chown -R obscura:obscura .

USER obscura

COPY . .

EXPOSE 8080

ENTRYPOINT [ "/usr/local/bin/python3" ]

CMD [ "main.py" ]
```

Maddosaurus commented:
That's a good idea. Actually, I'm currently thinking about switching this to a mutistage build (see [Solution 2 here](https://pythonspeed.com/articles/multi-stage-docker-python/) if we re-do the Dockerfile anyways.
This way, we can get a smaller and safer runtime image with less stuff installed.  
I'm currently playing around with this (thanks to this PR :) )

Maddosaurus commented:
I've ended with something like this, what do you think?
```
#
# Intermediate Compile Image
#

FROM python:3-alpine AS compile-image

RUN apk add --no-cache build-base

RUN python -m venv /opt/venv

# Basically this is everything import <venv>/bin/activate does
ENV PATH="/opt/venv/bin:$PATH"

COPY requirements.txt .
RUN pip install -r requirements.txt


#
# Final Build Image
#

FROM python:3-alpine AS build-image
COPY --from=compile-image /opt/venv /opt/venv

WORKDIR /cameraobscura

RUN apk add --no-cache shadow; \
    groupadd -r obscura && useradd -r -g obscura -d /cameraobscura -s /sbin/nologin -c "Docker image user" obscura; \
    chown -R obscura:obscura .

USER obscura

COPY . .

EXPOSE 8080

ENTRYPOINT [ "/opt/venv/bin/python3" ]

CMD [ "main.py" ]
```

m1k1o commented:
That is even better! That makes it down from 254.5 MB to 63.6 MB, what is about 85% saved space. Significant improvement.

Maddosaurus commented:
I've contacted the other maintainer and he'll take a look later today regarding possible implications of these changes.
If he is okay with them: Would you like to commit the multistage Dockerfile to this PR yourself or are you okay with me pushing these changes to your PR?
(I'm fine with giving you all the credit here as your PR set things in motion that led to this 💟  )

m1k1o commented:
I am ok with pushing your changes here. This is result of collaboration, and putting more heads together led to better and efficient results.

cmllr commented:
Hey, thank you for your contribution @m1k1o, looks good to me 👍

Maddosaurus committed:
As discussed: Introduce multistage Docker build

- Splitted into compile (venv) and build image
- Remove shadow dependency after useradd (improve security)

Maddosaurus commented:
I've gone ahead and pushed the multistage commit to your PR (thanks again!).  
Only thing I've changed additionally is to remove `shadow` after it served its purpose.  
Alpine actually doesn't include `useradd`/`groupadd` commands, hence the `shadow` package needs to be installed for that.  
Unfortunately, `shadow` [installs quite a lot of dangerous bins](https://pkgs.alpinelinux.org/contents?branch=edge&name=shadow&arch=x86&repo=community), therefore I feel better to just remove it after the fact :) 

Finally, your PR sparked discussions between multiple maintainers and might lead to a revival of this project 🎉

m1k1o commented:
That is a good caution, especially for a honeypot project. Maybe, in the future of this or any other project, swtichting to golang (or rust, any compiled language) and creating honeypot as a single executable put in `FROM scratch` image could even reduce attack surface.

> might lead to a revival of this project 🎉

Glad to hear that!
